src/rpc/reflect/parser.rs

```diff
@@ -0,0 +1,359 @@
+// Copyright 2019-2024 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+use std::collections::VecDeque;
+
+use serde::Deserialize;
+use serde_json::{json, Value};
+
+use super::{jsonrpc_types::RequestParameters, openrpc_types::ParamStructure, util::Optional as _};
+use crate::rpc::error::JsonRpcError;
+
+/// Parser for JSON-RPC parameters.
+/// Abstracts calling convention, checks for unexpected params etc.
+///
+/// Note that this type can panic on drop.
+#[derive(Debug)]
+pub struct Parser<'a> {
+    params: Option<ParserInner>,
+    /// What arguments do we expect to parse?
+    argument_names: &'a [&'a str],
+    /// How many times has the user called us so far?
+    call_count: usize,
+    has_errored: bool,
+}
+
+#[derive(Debug)]
+enum ParserInner {
+    ByPosition(VecDeque<Value>), // for O(1) pop_front
+    ByName(serde_json::Map<String, Value>),
+}
+
+impl Drop for Parser<'_> {
+    fn drop(&mut self) {
+        if !std::thread::panicking() && !self.has_errored {
+            assert!(
+                self.call_count >= self.argument_names.len(),
+                "`Parser` has unhandled parameters - did you forget to call `parse`?"
+            );
+        }
+    }
+}
+
+impl<'a> Parser<'a> {
+    pub fn new(
+        params: Option<RequestParameters>,
+        names: &'a [&'a str], // in position order
+        calling_convention: ParamStructure,
+    ) -> Result<Self, JsonRpcError> {
+        Self::_new(params, names, calling_convention).map_err(Into::into)
+    }
+    fn _new(
+        params: Option<RequestParameters>,
+        names: &'a [&'a str],
+        calling_convention: ParamStructure,
+    ) -> Result<Self, ParseError> {
+        let params = match (params, calling_convention) {
+            // ignore the calling convention if there are no arguments to parse
+            (None, _) => None,
+            (Some(params), _) if names.is_empty() && params.is_empty() => None,
+            // mutually exclusive
+            (Some(RequestParameters::ByPosition(_)), ParamStructure::ByName) => {
+                return Err(ParseError::MustBeNamed)
+            }
+            (Some(RequestParameters::ByName(_)), ParamStructure::ByPosition) => {
+                return Err(ParseError::MustBePositional)
+            }
+            // `parse` won't be called, so do additional checks here
+            (Some(RequestParameters::ByPosition(it)), _) if names.is_empty() && !it.is_empty() => {
+                return Err(ParseError::UnexpectedPositional(it.len()))
+            }
+            (Some(RequestParameters::ByName(it)), _) if names.is_empty() && !it.is_empty() => {
```

Comment:
let's have all the patterns documented. Some of them are obvious, some of them take a little while to understand.